compiler/rustc_codegen_cranelift/src/base.rs

```diff
@@ -513,6 +502,48 @@ fn codegen_fn_body(fx: &mut FunctionCx<'_, '_, '_>, start_block: Block) {
     }
 }
 
+fn codegen_alignment_check<'tcx>(
+    fx: &mut FunctionCx<'_, '_, 'tcx>,
+    pointer: mir::Operand<'tcx>,
+    required_alignment: u64,
+    source_info: mir::SourceInfo,
+) {
+    // Compute the alignment mask
+    let required_alignment = required_alignment as i64;
+    let mask = fx.bcx.ins().iconst(fx.pointer_type, required_alignment - 1);
+    let required = fx.bcx.ins().iconst(fx.pointer_type, required_alignment);
+
+    // And the pointer with the mask
+    let pointer = codegen_operand(fx, &pointer);
+    let pointer = match pointer.layout().abi {
+        Abi::Scalar(_) => pointer.load_scalar(fx),
+        Abi::ScalarPair(..) => pointer.load_scalar_pair(fx).0,
+        _ => unreachable!(),
+    };
+    let masked = fx.bcx.ins().band(pointer, mask);
+
+    // Branch on whether the masked value is zero
+    let is_zero = fx.bcx.ins().icmp_imm(IntCC::Equal, masked, 0);
+
+    // Create destination blocks, branching on is_zero
+    let panic = fx.bcx.create_block();
+    let success = fx.bcx.create_block();
+    fx.bcx.ins().brif(is_zero, success, &[], panic, &[]);
+
+    // Switch to the failure block and codegen a call to the panic intrinsic
+    fx.bcx.switch_to_block(panic);
+    let location = fx.get_caller_location(source_info).load_scalar(fx);
+    codegen_panic_inner(
+        fx,
+        rustc_hir::LangItem::PanicMisalignedPointerDereference,
+        &[required, pointer, location],
+        Some(source_info.span),
+    );
+
+    // Continue codegen in the success block
+    fx.bcx.switch_to_block(success);
```

Comment:
Please add a nop in the success block. This is necessary for clif ir printing annotations to work correctly.

Reply:
Why don't we need a nop in the panic block?

Reply:
It may be necessary there too.